Mars fetch functions

jonbry opened the pull request:
Updated `marsFetchRainfallData` and `marsFetchBaroData` to work with new dtime fields in the database. The functions still live in R/mars_data_fetch_write_functions.R (updated) with the old version living in either R/old_baro.R or R/old_rainfall.R

There are corresponding test files in the PWDGSI project folder, which show that it works both in March and November. These files plus the old_baro.R files will be eventually archived or deleted.

One thing to note, the old functions pull an extra day where the new ones don't. Those will show up on the diff tables

Closes #17 
Closes #19

jonbry committed:
Updated marsFetchRainfallData for new datetime

Produces the same results as previous function but works with dtime rather than dtime_est in the database. Removed zeros code because it is broken and will take more time to fix and marsFetchRainfallData is going to be superseded.

jonbry committed:
Old baro and rainfall functions

Used for comparison to make sure the new versions are working properly

jonbry committed:
Added testers

Tests old functions vs new

jonbry committed:
Updated test for november change

jonbry committed:
Updated baro function for new dtime

Did not re-add 5min interval

jonbry committed:
Removed sp from description

The sp package has a lot of hefty dependencies like terra, which isn't actively being used by PWDGSI functions

jonbry committed:
Updated notes and test functions

jonbry committed:
started dtime update for marsFetchLevel

jonbry committed:
Finished marsFetchWaterLevel

Close #22

jonbry commented:
@monica-water Can you also check `marsFetchLevelData` in this PR as well? I just added it since the other functions are pending.

Closes #22

jonbry committed:
Started marsFetchEvents to work with dtime in db

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -94,44 +92,46 @@ marsFetchRainfallData <- function(con, target_id, source = c("gage", "radar"), s
     stop("There is no data in the database for this date range.")
   }
-  
+  # Make sure rainfall is numeric?
```
I'm not sure why this variable cast was done originally. It's possible that the data was originally stored as text, but it's equally possible that this was from some old copy-pasted code that was needed in its original context. You can delete this - the data should always come in as numeric.

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -141,81 +141,82 @@ marsFetchRainfallData <- function(con, target_id, source = c("gage", "radar"), s
+  #     
+  #   }
+  # }
+  #### No including because this causes more issues than it should
```
So long as equivalent zero-punctuation is performed in the rainfall plotting function later, this is fine to not include here.

jonbry replied:
This currently does not work properly, but it can be added to the plotting function.

monica-water replied:
That makes sense. When it's time to revisit those functions, we can bring it up then.

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -141,81 +141,82 @@ marsFetchRainfallData <- function(con, target_id, source = c("gage", "radar"), s
+  # # }
+  # 
+  # return(finalseries)
 }
```
Is there any particular reason you're commenting the old code instead of just deleting it? We'll have archived copies in the source control backlog.

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -94,44 +92,46 @@ marsFetchRainfallData <- function(con, target_id, source = c("gage", "radar"), s
-  # rain_temp %<>% dplyr::filter(!is.na(dtime_edt)) %>% dplyr::arrange(dtime_edt)
-  rain_temp %<>% dplyr::filter(!is.na(dtime_est)) %>% dplyr::arrange(dtime_est)
+  # Return rainfall
+  rain_temp
```
Minor style nitpick: I prefer to call return() explicitly on the thing being returned. It helps communicate the finality of the function

jonbry replied:
This is a major debate with the R-core community on whether or not we should use `return()`. The primitive function `function()` returns the last value so using `return()` just means `function()` returns another function `return()`. We also describe the return obj/value in the roxygen comments. 

With that said, if this is something that's important to you, I can go back and add it in.

monica-water replied:
Let's not use it for now and I'll see if it makes a substantial difference for me in terms of code clarity. If it does, I'll bring it back up later.

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -370,77 +365,75 @@ marsFetchBaroData <- function(con, target_id, start_date, end_date, data_interva
+  baro <- DBI::dbGetQuery(con, paste0("SELECT * FROM data.viw_barodata_smp WHERE dtime >= '", start_date, "'", " AND dtime <= '", end_date + lubridate::days(1), "' order by dtime"))
+  #
+  baro_latest_dtime <- DBI::dbGetQuery(con, paste0("SELECT max(dtime) FROM data.tbl_baro WHERE dtime < '", end_date + lubridate::days(1), "'")) %>% dplyr::pull()
+  baro_latest_valid <- DBI::dbGetQuery(con, paste0("SELECT max(dtime) FROM data.viw_barodata_neighbors WHERE neighbors >= 4 and dtime < '", end_date + lubridate::days(1), "'")) %>% dplyr::pull()
```
Changing the threshold for the minimum number of neighbors required to return a composite time series has occasionally been discussed in the Analysis team group chat. Perhaps it's worth setting this as a named constant towards the top of the function for ease of changing in the future? Being hardcoded in the query makes it harder to find.

jonbry replied:
We are not doing any changes in this pass. This is just making sure the new dtime values work correctly. Future functions will not use hard coded queries

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -370,77 +365,75 @@ marsFetchBaroData <- function(con, target_id, start_date, end_date, data_interva
     #End value is 10 minutes after the final period because that 15 minute data point is good for 10 more minutes
-    baro_pad <- padr::pad(baro, start_val = min(baro$dtime_est), end_val = max(baro$dtime_est) + lubridate::minutes(10), interval = "5 mins")
-    
+    baro_pad <- padr::pad(baro, start_val = min(baro$dtimet), end_val = max(baro$dtime) + lubridate::minutes(10), interval = "5 mins")
```
dtimet appears to be a typo here

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -449,105 +442,105 @@ marsFetchBaroData <- function(con, target_id, start_date, end_date, data_interva
-  # browseURL(new_filename)
-  
-  #return Final Series. 
-  return(finalseries)
```
I prefer an explicit call to return() for clarity's sake

monica-water replied:
Changing my mind on this!

monica-water commented on R/mars_data_fetch_write_functions.R:
```diff
@@ -852,48 +845,48 @@ marsFetchSMPSnapshot <- function(con, smp_id, ow_suffix, request_date){
+  # Select query table
+  #### This needs to be reworked/simplified with str_detect and only one else if
+  #### Can we get rid of PZ because it doesnt' seem to exist in ow_prefixes
+  #### This replace gets rid of the number at the end. It should be just letters before number?
```
The str_replace strips the terminal number because only the first two letters are what determines what kind of data is being sampled at the well.

PZ (Piezometer) seems to have been renamed internally since this code was written - there are no PZ wells in fieldwork.tbl_ow. It can be removed from this code.

I'll bring up OW naming with Billy to make sure ow_prefixes is up to date,

monica-water reviewed (approved):
I have a few minor stylistic requests (deleting large old commented code blocks and explicit calls to return() when functions return their final output).

There seems to be a typo on line 421 of fetch_write_functions.R

Otherwise, this looks great!

jonbry committed:
Deleted commented out code

Per Monica

jonbry committed:
Pulled changes from master

Merge branch 'master' of https://github.com/PWD-MARS/pwdgsi into mars-fetch

# Conflicts:
#	R/old_events.R

jonbry committed:
updated old_events

jonbry committed:
Verfied diffs in db are ok for FetchRainEvents

Using FetchRainEvents gives two different values because the values in the DB are different.The issue stems from the maintenance script that writes to the DB which didn't handle 00:00:00 observations. Verified this was the issue

Closes #16

jonbry committed:
Updated marsFetchMonitoringData for dtime

Updated to work with new dtime structure in the database as well as comments so they reflect the purpose and output of the function

Closes #31

jonbry committed:
Verified yday_decimal works with datetime

jonbry committed:
started updating snapshot

They are listed as having dtime_est but the variables in the db have not been updated so Monica is looking into them
